Further contract cleanup and tests

andresilva opened the pull request:
Depends on #3.

- Addressed grumbles from PR #2
  - Enabled all solium lints (disabled mixedcase checking for variables so we don't break the ABI)
  - All parameter lists longer than 3 are broken (this is checked by solium)
  - All function modifiers are declared in a new line (one per line)
  - Documented behavior of `addRelease` and `addChecksum` functions
  - Added extra check on all functions that change client ownership (the new owner shouldn't already be an owner of a client)
  - Added check for client existence when adding it
  - Test new behavior
- SimpleOperations
  - Simplify adding/setting/removing a client
  - Remove fallback function
  - Validate release exists when adding checksum 
- OperationsProxy
  - Remove support for arbitrary transaction relaying
  - ~~Remove support for automatic confirmation of requests (if no confirmer defined)~~
  - Removed payable functions and contract kill
  - Clean up release state variables automatically
  - Add tests

andresilva committed:
enable more solium lints

andresilva committed:
break large parameter lists and move function modifiers one per line

andresilva committed:
update copyright date

andresilva committed:
check new owner isn't already an owner when moving client ownership

andresilva committed:
document behavior of addRelease and addChecksum functions

andresilva committed:
check new client doesn't already exist when adding it

andresilva committed:
SimpleOperations: test check for adding duplicate client

andresilva committed:
add test for checking that an owner doesn't own multiple clients

andresilva committed:
fix async assert of function throw

andresilva committed:
OperationsProxy: test constructor

andresilva committed:
OperationsProxy: add tests for setting owner, delegate and confirmer

andresilva committed:
OperationsProxy: require a confirmer for all tracks

tomusdrw commented on contracts/OperationsProxy.sol:
```diff
@@ -49,80 +61,121 @@ contract OperationsProxy {
 
-	function() public only_owner {
+	function()
+		public
```
`relay()` is using `msg.value` I suspect it should be `payable` same as this function.

andresilva replied:
I think both `OperationsProxy` and `SimpleOperations` should have no `payable` functions since they don't do anything with it.

tomusdrw commented on contracts/OperationsProxy.sol:
```diff
@@ -49,80 +61,121 @@ contract OperationsProxy {
 		delete trackOfPendingRelease[_release];
 	}
 
 	function kill() public only_owner {
```
Can we get rid of this? Let's call it `withdraw` and just send the funds without deleting contract code.

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,106 +46,210 @@ contract SimpleOperations is Operations {
+		public
+		only_owner
+	{
 		resetClientOwner(_client, 0);
```
This won't work, after this function is called `0` becomes an owner of `_client` and then you cannot remove any other client.

tomusdrw replied:
Why do we need so many events and functions anyway? Why not just `setClient(_client, _owner)` (covers adding and changing) and `removeClient(_client, _owner)` that get's rid of all traces of particular client (TBH I'm not sure if it's even needed after we removed the fork voting stuff)?

tomusdrw commented on test/simple_operations.js:
```diff
@@ -1,15 +1,420 @@
 
 contract("SimpleOperations", accounts => {
+  const assertThrowsAsync = async (fn, matcher) => {
+    let f = () => {};
```
Why not just:
```js
try {
  await fn();
} catch (e) {
  assert.error(e, matcher);  // don't know the syntax here
  return;
}

assert.fail('Expected fn to throw');
```

tomusdrw commented on test/simple_operations.js:
```diff
@@ -1,15 +1,420 @@
+    // we can't transfer ownership of the parity-light client to `accounts[0]` since it is already
+    // an owner of the parity client
+    await assertThrowsAsync(
+      async () => await operations.setClientOwner(accounts[0], { from: accounts[1] }),
```
you can drop `async` and `await` since it just returns a promise anyway

andresilva committed:
simplify assertThrowsAsync

andresilva committed:
remove fallback function from SimpleOperations

andresilva committed:
SimpleOperations: simplify adding/setting client owner

andresilva committed:
OperationsProxy: remove support for relaying arbitrary transactions

andresilva committed:
OperationsProxy: remove payable modifiers and contract suicide

andresilva committed:
OperationsProxy: include request status in request confirmed event

andresilva committed:
OperationsProxy: clean up release automatically

andresilva committed:
SimpleOperations: require release exists when adding checksum

andresilva committed:
use mixed case for modifiers

andresilva committed:
OperationsProxy: test fallback function

andresilva committed:
OperationsProxy: test adding requests and confirming them

andresilva committed:
OperationsProxy: test rejection of request

andresilva committed:
OperationsProxy: test addChecksum fetching track data from contract

andresilva committed:
OperationsProxy: test validate track confirmer exists

andresilva committed:
replace deprecated var keyword with explicit type

andresilva committed:
Operations: remove unused state variables

andresilva committed:
OperationsProxy: fix possible re-entrancy in confirm

andresilva committed:
add braces to all if statements

andresilva committed:
update dependencies and force truffle to use solc 0.4.21

andresilva committed:
update to solc 0.4.21 and use emit keyword

andresilva commented:
I updated to solc 0.4.21 and started using the `emit` keyword since I added support for it in `solidity-coverage` (https://github.com/sc-forks/solidity-parser/pull/13) and emacs (https://github.com/ethereum/emacs-solidity/pull/24). :bowtie:

kirushik reviewed (approved):
Looks good to me.
I still want to suggest using `yarn` for dependencies instead of pure `npm` — people who would have several contracts checked out side-by-side will thank you later;-)

andresilva committed:
SimpleOperations: don't leak information from removed clients

andresilva committed:
OperationsProxy: relay requests right away if no confirmer is defined

tomusdrw commented on contracts/OperationsProxy.sol:
```diff
@@ -49,95 +61,147 @@ contract OperationsProxy {
-		if (relayOrConfirm(_track))
-			AddReleaseRelayed(_track, _release);
-		else
+	function addRelease(
```
Do you mind adding an additional modifier here to make sure it can only be called by `onlyDelegateFor(_track)`?
Currently it will fail calling `tryRelay` function, but imho it would be better to actually duplicate it here to make it easier to analyse.

andresilva replied:
Agreed, I will also do the same thing for `addChecksum`. Maybe the optimizer even removes the redundant check.

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,121 +46,258 @@ contract SimpleOperations is Operations {
+		returns (bool)
+	{
+		Release storage r = client[_client].release[_release];
+		return r.forkBlock != 0 && r.track != 0 && r.semver != 0;
```
Seems that we require forkBlock to be always set to a non-zero value. I don't think it's intuitive, better to ditch that check imho.

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,121 +46,258 @@ contract SimpleOperations is Operations {
+		view
+		returns (bool)
+	{
+		return client[_client].owner != 0;
```
This is a bit tricky, seems that setting a new owner of a client to `0x0` (which is not forbidden) will cause the client to be treated as removed

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,121 +46,258 @@ contract SimpleOperations is Operations {
 	}
 
-	function clientOwner(address _owner) public view returns (bytes32) {
+	function clientOwner(address _owner)
```
Function name is a bit misleading imho, `clientExists` takes `_client` and returns true/false, `clientOwner` however takes `_owner` and returns a client name. Why do we need a function for this anyway?

andresilva replied:
I agree that the function name is misleading, it's a carryover from the naming of the state variables. IMO `client` and `clientOwner` should have their names switched with one another. This method is defined in the interface, and the `OperationsProxy` uses it to get the name of the client it is managing.

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,121 +46,258 @@ contract SimpleOperations is Operations {
+		view
+		returns (bytes32)
+	{
+		if (clientExists(_client)) {
```
Why not just `require(clientExists(_client))`?

andresilva replied:
I thought it would be better to keep the existing interface (similar to returning `None`) rather than abort.

tomusdrw replied:
ok, it's fine then.

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,121 +46,258 @@ contract SimpleOperations is Operations {
+		view
+		returns (bytes32 o_release, bytes32 o_platform)
+	{
+		if (clientExists(_client)) {
```
Same as above

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,121 +46,258 @@ contract SimpleOperations is Operations {
-	function setClientOwner(address _newOwner) public only_client_owner {
-		var newClient = clientOwner[msg.sender];
-		clientOwner[msg.sender] = bytes32(0);
+	function setClientOwner(address _newOwner)
```
Setting a client owner to `0x0` cause the client to be treated as removed (see comment in `clientExists()`). Best to disallow setting it to `0x0`

tomusdrw commented on contracts/SimpleOperations.sol:
```diff
@@ -45,121 +46,258 @@ contract SimpleOperations is Operations {
+		onlyOwner
+		notClientOwner(_owner)
+	{
+		address owner = client[_client].owner;
```
Same here, we should disallow setting `0x0` as an owner, the owner should just use `removeClient` instead.

tomusdrw reviewed (commented):
Couple of grumbles.

andresilva committed:
OperationsProxy: move input validation modifiers to public methods

andresilva committed:
OperationsProxy: merge tryRelease and addRequest

andresilva committed:
SimpleOperations: validate non-zero address when setting client owner

andresilva committed:
SimpleOperations: don't require release forkBlock to be non-zero

andresilva committed:
SimpleOperations: require release track and semver are non-zero
